scipy/interpolate/_ppoly.pyx

```diff
@@ -458,6 +458,11 @@ def real_roots(double[:,:,::1] c, double[::1] x, double y, bint report_discont,
 
     wr = <double*>libc.stdlib.malloc(c.shape[0] * sizeof(double))
     wi = <double*>libc.stdlib.malloc(c.shape[0] * sizeof(double))
+    if not wr or not wi:
+        libc.stdlib.free(wr)
+        libc.stdlib.free(wr)
+        raise MemoryError("Failed to allocate memory in real_roots")
```

Comment:
@WarrenWeckesser You left comments in https://github.com/scipy/scipy/pull/10035/files#r279160979 about `raise` not working in `cdef` without `except -1`. I read you link and the way I processed it is that `def` functions are pure python and I'm allowed to `raise` like normal in them.

I believe I have correctly added `except -1` or similar where needed adding a return int to some functions. I'm still new to cython so a more experienced pair of eyes would be appreciated. Thanks